### PR TITLE
PERF: avoid copy of dataframe while writing to SQL DBs

### PR DESCRIPTION
**Pull request by sridhu88**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] This change avoids making an entire copy of the dataframe while writing to SQL DBs with a chunksize. Instead we convert slices of the dataframe into a data list and write to the DB. Saves significant memory when writing large dataframes to SQL DBs.

### CODE REVIEW

File: pandas/io/sql.py [hunk 1]
```diff
@@ -608,17 +608,25 @@ def _execute_insert_multi(self, conn, keys, data_iter):
...
 
-    def insert_data(self):
+    def insert_data(self, start_index, end_index):
+        """
+        Creates list of col vectors for a slice of the dataframe.
+        Slice range specified by start_index, end_index
+        """
         if self.index is not None:
-            temp = self.frame.copy()
+            temp = self.frame[start_index:end_index].copy()
```
Comment (jreback) [inline pandas/io/sql.py]: these should all be .iloc

File: pandas/io/sql.py [hunk 2]
```diff
@@ -608,17 +608,25 @@ def _execute_insert_multi(self, conn, keys, data_iter):
...
             temp.index.names = self.index
             try:
                 temp.reset_index(inplace=True)
             except ValueError as err:
                 raise ValueError(
                     "duplicate name in index/columns: {0}".format(err))
         else:
-            temp = self.frame
+            if start_index == 0 and end_index == len(self.frame.index):
```
Comment (jreback) [inline pandas/io/sql.py]: this shouldn't be necessary

**Review by jreback — CHANGES_REQUESTED** [review]
you would need to test this

**Comment by sridhu88** [comment]
> you would need to test this

Sorry I'm new to Pandas development. Are you suggesting adding a test to demonstrate performance or just the correctness? If correctness, I thought there are already quite a few tests for the to_sql function which this change affects.

**Comment by jreback** [comment]
@sridhu88 a performance test that this actually improves things, see the asv_bench dir

**Comment by jorisvandenbossche** [comment]
Jeff, I am not sure it is easy to test this.  There are already sql tests in ASV, so we can run them and see if we see a difference, but I am not sure we will (copying is relatively cheap compared to the writing to sql). And we don't test memory in ASV.

**Comment by jreback** [comment]
@jorisvandenbossche  https://github.com/pandas-dev/pandas/pull/25255#issuecomment-463202722

I agree, which is why I don't think this PR actually helps any, but I could be wrong.

**Comment by jorisvandenbossche** [comment]
Actually, a specific benchmark that uses `chunksize` might be good to add (just checked and that is something we don't yet have) *in general*.

I quickly did a test with/without chunksize on master and this PR, and this PR is actually slower. Which is somewhat to be expected. We now chunk the dataframe before converting to arrays, while before it was slicing the arrays (and slicing arrays is faster than slicing dataframes). I still think this PR can be worth it, as it is a trade-off between memory and performance.

Howeve

**Comment by jorisvandenbossche** [comment]
(Jeff, didn't see your comment before posting the last one, github is sometimes annoying :-))

@sridhu88 do you have a benchmark that measures memory that shows the impact of this PR? 
Without necessarily including that as a test, being able to show that is probably essential for deciding whether we want to merge this or not.

**Comment by sridhu88** [comment]
Thanks Jeff, Joris. Just to clarify, I labeled this PR as "PERF" purely for memory reasons. I was seeing that to insert a dataframe of ~2G, the peak memory during to_sql is approximately 2-3 times the dataframe size. With this change, the peak memory will remain very close to the actual size of the dataframe. As you mentioned, there may be a small hit to runtime. But in my tests, I didn't see a significant slowdown. Let me set up some benchmarking tests and get back to you soon.

**Comment by WillAyd** [comment]
Is this not something we can measure with ASV?

https://asv.readthedocs.io/en/stable/writing_benchmarks.html#memory

**Comment by WillAyd** [comment]
@sridhu88 can you see if we can get the ASV running and merge master?

**Comment by WillAyd** [comment]
Closing as stale. Ping if you'd like to continue

**Pull request closed without merging**